### PR TITLE
Show More button in VJ Includes on Amp

### LINKED ISSUE #6841: Render a Show More button in VJ Includes on AMP
### Is your feature request related to a problem? Please describe.
Following the completion of https://github.com/bbc/simorgh/issues/6917, we will be able to render VJ Includes on AMP. We need to add a Show More button to these.

### Describe the solution you'd like

Using the `resizable` attribute on `amp-iframe` we should implement a Show More button - see example: https://amp.dev/documentation/components/amp-iframe/#iframe-resizing.

#### Translations
"Show More" will have translations. In this issue, we will implement the logic to retrieve this translation from the `showMore` property, but default to "Show More" if that's not available. (Translations will be added in a separate issue).

#### Designs
When integrated successfully with the include, the 'show more' button that appears on VJ includes looks like this:
image

#### Note

VJ Includes on AMP already include the postMessage code to support this behaviour.
```
// https://news.files.bbci.co.uk/include/vjamericas/176-eclipse-lookup/mundo/app/amp
resizeSetup: function () {
  setInterval(function() {
    window.parent.postMessage({
      sentinel: 'amp',
      type: 'embed-size',
      height: document.body.scrollHeight
    }, '*');
  }, 500);
},
```

### Describe alternatives you've considered
n/a

### Testing notes
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [x] This feature is expected to need manual testing.
- [x] Cypress tests would need to be created to check the 'show more' button is rendering
- [x] Needs an a11y review with Catharine during development

### Additional context
n/a

### PR DESCRIPTION
**Pull request by Simon2828**
Resolves #6841

**Overall change:**
Add a show more button to VJ includes on AMP which are bigger than the placeholder dimensions.

**Code changes:**

- Add `resizable` attribute to `amp-iframe`
- Include an overflow element and button
- Style using the [pocket guidelines](https://www.bbc.co.uk/gel/guidelines/pocket#overview)

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] I have added labels to this PR for the relevant pod(s) affected by these changes
- [X] I have assigned this PR to the Simorgh project

**Testing:**

- [X] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [X] This PR requires manual testing

---

**Dev/A11y Swarm**

Test asset: http://localhost:7080/mundo/23263889.amp
⏳ Needs a11y laptop.

https://bbc.github.io/accessibility-news-and-you/accessibility-news-and-testers

- [x] Axe
- [x] Design
  - Needs :focus and :hover styles linked.
- [x] Colour Contrast
- [x] Colour blindness
- [ ] Windows HCM ⏳
- [x] FF Colour Preferences
  - Resolve button not inheriting new colour scheme:
  - Screenshot 2020-07-14 at 12 29 17

- [x] Text 200%, Page 200%
- [ ] ~~Icons~~
- [ ] ~~Heading Order~~
- [x] Check accessibility of browser DOM
- [x] Keyboard Navigation 

**_Screen Readers (IE11)_** 
- [ ] ZoomText Magnifier/Reader ⏳
- [ ] Dragon NaturallySpeaking ⏳
- [ ] JAWS ⏳
- [ ] Read&Write ⏳
- [ ] NVDA (Firefox) ⏳

**_Mac, iOS, and Android:_** 
- [x] VoiceOver iOS (iPhone/iPad)
- [x] VoiceOver OS (Mac)
- [x] TalkBack (Chrome on Android)

### CODE REVIEW

File: data/mundo/cpsAssets/vj-23263889.json [hunk 1]
```diff
@@ -0,0 +1,678 @@
...
+          {
+            "name": "Related URL Link 2",
+            "uri": "http://www.bbc.co.uk/news",
+            "type": "link"
+          }
+        ]
+      }
+    ]
+  }
+}
```
Comment (andrewscfc) [inline data/mundo/cpsAssets/vj-23263889.json]: Where did this test asset come from? I've never seen one prefixed with `vj` before?

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 2]
```diff
@@ -7,6 +7,39 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
   display: grid;
 `;
 
+const PocketTapArea = styled.div`
+  &&& {
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx]: Looked this up here, is there a reason you need override styles: https://styled-components.com/docs/faqs#how-can-i-override-styles-with-higher-specificity ?

I wouldn't expect us to need to do this as I'd expect styles to be encapsulated as we are using CSS in JS.

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 3]
```diff
@@ -7,6 +7,39 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+    display: block;
+    height: 70px;
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx]: I'm not entirely sure if it's written anywhere but we should always use REM in our codebase so the height responds to the browser font size choice for accessibility reasons.

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 4]
```diff
@@ -7,6 +7,39 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+      rgba(255, 255, 255, 0.816947) 66%,
+      rgba(255, 255, 255, 0.0718488) 100%
+    );
+    &:after {
+      display: inline-block;
+      content: '';
+      width: 100vw;
+      border-bottom: 2px solid #1380a1;
+      transform: translateY(2.4rem);
+    }
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx:67]: Where did these styles come from are they ported from legacy amp implementation? Have we confirmed with UX that this styling is what we want?

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 5]
```diff
@@ -7,6 +7,39 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+  }
+`;
+
+const PocketButton = styled.button`
+  background: #1380a1;
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx]: Once again re porting these styles from somewhere else, I think the colours should generally come from our colours in psammead, these have been agreed with UX: https://github.com/bbc/psammead/blob/latest/packages/utilities/psammead-styles/src/colours.js

**Review by andrewscfc — COMMENTED** [review]
Great first pass at this 👍 

I think `PocketTapArea` and `PocketButton` should be moved to psammead as they are self-contained UI components. As I've mentioned in some of my comments, these components will need a UX review particularly if the styles have been ported from the legacy AMP site.

File: data/mundo/cpsAssets/vj-23263889.json (hunk 1 above)
Comment (Simon2828) [reply data/mundo/cpsAssets/vj-23263889.json]: I created this asset to add VJs which have the show more button displayed on them, as they are bigger than their placeholder dimensions. Would it be better to add a bigger VJ to the mundo asset we normally use instead?

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 2 above)
Comment (Simon2828) [reply src/app/containers/Include/amp/VjAmp.jsx]: I used this to override the `i-amphtml-element>[overflow]` attribute selector which Amp adds to the amp-iframe. The attribute selector in the amp-iframe seems to be injected after the styled-component, so overrides the style, unless the ampersands are used to make the PocketTapArea styles more specific.

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 4 above)
Comment (Simon2828) [reply src/app/containers/Include/amp/VjAmp.jsx:67]: UX do not have designs on the show more button, but yes am in conversation with them to get the stylings they want.

**Comment by Simon2828** [comment]
> Great first pass at this 👍
> 
> I think `PocketTapArea` and `PocketButton` should be moved to psammead as they are self-contained UI components. As I've mentioned in some of my comments, these components will need a UX review particularly if the styles have been ported from the legacy AMP site.

Thanks @andrewscfc , yes agree that we should make a psammead component for this, but would prefer to raise a new issue for this. Currently the `PocketTapArea` and `PocketButton` are rendered in the `amp-iframe` and think it would take more work to make a psammead component which works independently of the `amp-iframe`. Given this issue should be completed soon, rather than switching context now, is it ok to carry on with the changes in simorgh, then raise the psammead component issue afterwards?

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 2 above)
Comment (Simon2828) [reply src/app/containers/Include/amp/VjAmp.jsx]: @simonsinclair has showed me a cleaner approach 👍

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 6]
```diff
@@ -1,34 +1,96 @@
...
 import styled from 'styled-components';
-import { shape, string } from 'prop-types';
+import { shape, string, node } from 'prop-types';
+import {
+  GEL_SPACING,
+  GEL_SPACING_DBL,
+} from '@bbc/gel-foundations/dist/spacings';
 import { GridItemConstrainedMedium } from '#lib/styledGrid';
 
+const KEYLINE_WIDTH = '0.125rem';
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx:15]: What does 'KEYLINE' refer to is that a design term?

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 7]
```diff
@@ -1,34 +1,96 @@
...
+    sandbox="allow-scripts allow-same-origin allow-top-navigation-by-user-activation allow-forms"
+    resizable
+    src={src}
+  >
+    {children}
+  </amp-iframe>
+);
+
+const StyledAmpIframe = styled(AmpIframe)`
+  > [overflow] {
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx:46]: What does `[overflow]` do here, is that pure CSS or styled components?

**Review by andrewscfc — COMMENTED** [review]
Just checked this out and found a bug I think, on first load the 'Show More' button doesn't initially appear then eventually appears at the top of the include:
This is in firefox:
image

Also sometimes the Show More button never appears at all:
image
http://localhost:7080/mundo/23263889.amp

On a different note, I think we also need to implement 'Show Less' to match the legacy implementation and the GEL guidlines. I guess that could be a different issue if it's acceptable to product we don't implement 'Show Less' initially, though that doesn't sound a great UX.

**Comment by andrewscfc** [comment]
> is it ok to carry on with the changes in simorgh, then raise the psammead component issue afterwards 

Yeah that's absolutely fine, the styling is quite coupled to the amp-iframe so maybe that would have to go into psammead too; not sure if that's the right thing to do or not?

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 6 above)
Comment (Simon2828) [reply src/app/containers/Include/amp/VjAmp.jsx:15]: Yes, it refers to the blue horizontal line coming out of the show more button

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 7 above)
Comment (simonsinclair) [reply src/app/containers/Include/amp/VjAmp.jsx:46]: It's an attribute selector – https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Selectors/Attribute_selectors
It's used in the AMP stylesheet to style the overflow element and we use it to update the styles with the same specificity.

Comment (Simon2828) [reply src/app/containers/Include/amp/VjAmp.jsx:46]: Yes, it is pure CSS. It is selecting the overflow element where styles have been injected by the amp-iframe and then the styles set in `StyledAmpIframe` override these styles set by the amp-iframe. Previously I used the `&&&` which was increasing specificity by adding multiple classnames, however @simonsinclair showed me that using this child element selector `> [overflow]`, is a cleaner approach, as now we are overriding the styles on the element where they are set.

Comment (andrewscfc) [reply src/app/containers/Include/amp/VjAmp.jsx:46]: Thanks for explaining guys, makes sense 😄

**Comment by Simon2828** [comment]
> Just checked this out and found a bug I think, on first load the 'Show More' button doesn't initially appear then eventually appears at the top of the include:
> This is in firefox:
> image
> 
> Also sometimes the Show More button never appears at all:
> image
> http://localhost:7080/mundo/23263889.amp
> 

Was unable to reproduce this bug on first load, but could reproduce it when resizing the browser viewport quickly. When taking out the styles we had applied to the show more button, the bug is still reproducible. The `resize` event listener reappears on the overflow element and so the button displays again. Will discuss with UX whether this is enough of an edge case to be acceptable.

> On a different note, I think we also need to implement 'Show Less' to match the legacy implementation and the GEL guidlines. I guess that could be a different issue if it's acceptable to product we don't implement 'Show Less' initially, though that doesn't sound a great UX.

The 'Show Less' will be displayed in Includes which already contain this button, so will be matching the legacy implementation in these cases. However for Includes which do not have a 'Show Less', Amp does not give us the option to add one. The `amp-iframe` only has an overflow / show more element, which we are styling - we are not writing its functionality.

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 8]
```diff
@@ -18,31 +28,98 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+        <StyledAmpIframe
           width={imageWidth}
           height={imageHeight}
-          layout="responsive"
-          sandbox="allow-scripts allow-same-origin allow-top-navigation-by-user-activation allow-forms"
           src={src}
+          service={service}
         >
+          <div overflow="true">
+            <button type="button" aria-label="Show More">
```
Comment (simonsinclair) [inline src/app/containers/Include/amp/VjAmp.jsx]: We should probably update the case of the `aria-label` too.

**Review by simonsinclair — APPROVED** [review]
Fantastic work, Simon. Thanks for all the effort you've put into this.
Please see my comment below. Let's also arrange a final check-in with @nataliesmart.

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 9]
```diff
@@ -18,31 +28,98 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+      background-color: #fff;
+      border-top: ${KEYLINE_WIDTH} solid ${C_EBON};
+      content: '';
+      display: block;
+      height: 50%;
+      left: 0;
+      position: absolute;
+      top: 50%;
+      width: 100%;
+      z-index: -10;
```
Comment (andrewscfc) [inline src/app/containers/Include/amp/VjAmp.jsx:66]: Where does the `z-index` `-10` value come from, I guess it makes the`show more' button 'low priority', what is it competing with?

**Review by andrewscfc — APPROVED** [review]
Great work guys 🌮 

Now this is pretty much done, what do you guys think about moving some of this to `psammead`? Is that a good idea or is it a bit too special for that? As we said earlier in a different issue if we do it at all.

**Comment by Simon2828** [comment]
> Great work guys 🌮
> 
> Now this is pretty much done, what do you guys think about moving some of this to `psammead`? Is that a good idea or is it a bit too special for that? As we said earlier in a different issue if we do it at all.

Thanks @andrewscfc.  Yes, I created a new psammead issue https://github.com/bbc/psammead/issues/3597

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 9 above)
Comment (simonsinclair) [reply src/app/containers/Include/amp/VjAmp.jsx:66]: This rule is added to the pseudo-element we create with `::after`, which creates the 'keyline'. It ensures it's always _behind_ the button.
It's not competing with anything as such - I tend to follow the [popular](https://tailwindcss.com/docs/z-index) practice of rounding z-indexes to multiples of 10 as it gives you a bit of breathing room and makes them easier to keep track of.

**Comment by paruchurisilpa** [comment]
Tested this using the link : http://localhost:7080/mundo/23263889.amp

Noticed the following issues:

* Show more doesn't get displayed when you view the page first time, you have to refresh the page and the include has to be visible on the page for 'Show more' to appear
* In smaller viewports 'Show more' appears on the top of the include

Screenshot 2020-07-14 at 16 29 27

* Page is clickable anywhere ( Known issue)

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 10]
```diff
@@ -18,31 +28,98 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+    button {
+      ${({ service }) => getSansBold(service)}
+      ${GEL_PICA}
+      background-color: ${C_EBON};
+      border: none;
+      color: #fff;
+      cursor: pointer;
+      display: block;
+      padding: ${GEL_SPACING} ${GEL_SPACING_DBL};
+      &:hover {
```
Comment (greenc05) [inline src/app/containers/Include/amp/VjAmp.jsx]: missing focus styling?

File: src/app/containers/Include/amp/VjAmp.jsx [hunk 11]
```diff
@@ -18,31 +28,98 @@ const IncludeGrid = styled(GridItemConstrainedMedium)`
...
+        <StyledAmpIframe
           width={imageWidth}
           height={imageHeight}
-          layout="responsive"
-          sandbox="allow-scripts allow-same-origin allow-top-navigation-by-user-activation allow-forms"
           src={src}
+          service={service}
         >
+          <div overflow="true">
+            <button type="button" aria-label="Show more">
```
Comment (greenc05) [inline src/app/containers/Include/amp/VjAmp.jsx]: Please remove the aria-label, this is repeating the text inside the element...

Comment (greenc05) [inline src/app/containers/Include/amp/VjAmp.jsx]: do we need type button?

Comment (greenc05) [reply src/app/containers/Include/amp/VjAmp.jsx]: Unless we are actually going to say what we are going to show more of? Which we should... Do we have some functionality that ensures the content that is not shown is hidden from AT until this button is clicked? I'm not sure it's a very accessible implementation we currenlty have on live...

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 10 above)
Comment (simonsinclair) [reply src/app/containers/Include/amp/VjAmp.jsx]: Please see the a11y review results in the description.

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 11 above)
Comment (simonsinclair) [reply src/app/containers/Include/amp/VjAmp.jsx]: Don't forget this is AMP and we're bound to their specification: https://amp.dev/documentation/components/amp-iframe/#iframe-resizing. We can probably remove the `aria-label` here now that this is a `button` and not a `div` though, @Simon2828.

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 10 above)
Comment (greenc05) [reply src/app/containers/Include/amp/VjAmp.jsx]: Thank you, had not seen that

File: src/app/containers/Include/amp/VjAmp.jsx (hunk 11 above)
Comment (simonsinclair) [reply src/app/containers/Include/amp/VjAmp.jsx]: Yes, declaring the input type ensures it doesn't become a `submit` or `reset` button.

**Comment by greenc05** [comment]
FYI as AMP is mobile, there is no need to a11y swarm with the Windows AT:

- ZoomText Magnifier/Reader 
- Dragon NaturallySpeaking 
- JAWS 
- Read&Write 
- NVDA (Firefox) 
- Windows HCM

**Comment by greenc05** [comment]
@Simon2828 please could you add a transparent border to the button - this will help this element look more like a button when users change their colour preferences

**Comment by greenc05** [comment]
Please can we also test Voice Control on iPhone and Voice Access on Andriod @Simon2828

**Comment by greenc05** [comment]
@Simon2828 I see you've mentioned GEL 'pocket' guidelines, there is also this if you haven't seen it, not sure if it's useful https://bbc.github.io/gel/components/pockets/

**Comment by greenc05** [comment]
@Simon2828 FYI "Resolve button not inheriting new colour scheme" buttons don't inherit colours like other elements - bug with FF I believe

**Comment by Simon2828** [comment]
> Please can we also test Voice Control on iPhone and Voice Access on Andriod @Simon2828

Voice Access works fine on Android 👍 
Please could you test Voice Control on iphone @simonsinclair ?

**Comment by Simon2828** [comment]
> @Simon2828 FYI "Resolve button not inheriting new colour scheme" buttons don't inherit colours like other elements - bug with FF I believe

Thanks @greenc05 - I have tried many different ways to change the text of the button to show the updated colour scheme in FF but have been unable to get it working. Given the bug in FF, is it ok to leave the styling like this: https://user-images.githubusercontent.com/1430787/87420765-bee70480-c5cd-11ea-92ec-785ae3bcc46d.png ?

**Comment by simonsinclair** [comment]
> Voice Access works fine on Android 👍
> Please could you test Voice Control on iphone @simonsinclair ?

Tested, and also working on iOS.

**Comment by greenc05** [comment]
Yep @Simon2828 just needs transparent border adding if you haven't already please https://github.com/bbc/simorgh/pull/7171#issuecomment-658684039

**Pull request merged**